src/Infrastructure.Styles.Analyzer/RMSYN1103ExtensionMethodsTestedAsNormalInvocationCodeFixProvider.cs

```diff
@@ -0,0 +1,145 @@
+using System.Buffers;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Infrastructure.Styles.Analyzer
+{
+    public class RMSYN1103ExtensionMethodsTestedAsNormalInvocationCodeFixProvider : CodeFixProvider
+    {
+        public override FixAllProvider? GetFixAllProvider ()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public override Task RegisterCodeFixesAsync (CodeFixContext context)
+        {
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    "Fix Extension method invocation",
+                    c => CodeFix(c, context),
+                    RMSYN1103ExtensionMethodsTestedAsNormalInvocationSyntaxAnalyzer.DiagnosticId),
+                context.Diagnostics.First());
+            return Task.CompletedTask;
+        }
+
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } =
+            ImmutableArray.Create(RMSYN1103ExtensionMethodsTestedAsNormalInvocationSyntaxAnalyzer.DiagnosticId);
+
+        private static async Task<Document> CodeFix (CancellationToken cancellationToken, CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.First();
+            var diagnosticSpan = diagnostic.Location.SourceSpan;
+
+            var syntaxRoot = await context.Document.GetSyntaxRootAsync(cancellationToken);
+            if (syntaxRoot == null) return context.Document;
+
+            var oldNode = syntaxRoot.FindNode(diagnosticSpan).FirstAncestorOrSelf<InvocationExpressionSyntax>();
+            if (oldNode == null) return context.Document;
+
+            var semanticModel = await context.Document.GetSemanticModelAsync(cancellationToken);
+            if (semanticModel == null) return context.Document;
+
+            var symbolInfo = semanticModel.GetSymbolInfo(oldNode, cancellationToken);
+            if (symbolInfo.Symbol is not IMethodSymbol methodSymbol)
+                return context.Document;
+
+            var containingType = methodSymbol.ContainingType;
+            if (containingType == null) return context.Document;
+
+            var newSyntaxRoot = syntaxRoot.ReplaceNode(
+                oldNode, 
+                TransformExtensionInvocationNode(
+                    oldNode, 
+                    containingType, 
+                    semanticModel)
+                );
+            
+            return context.Document.WithSyntaxRoot(newSyntaxRoot);
+        }
+
+        private static SyntaxNode TransformExtensionInvocationNode (InvocationExpressionSyntax oldNode, INamedTypeSymbol containingType,
```

Comment:
This method seems overly complex for the task as it uses recursion and a lot of tracing. Maybe it is necessary but I want to understand why. Consider the example `` with an invocation ``. 

```c#
namespace This.That;

public static class Ext
{
  public static void TestX(this string a, int b) { ... }
}
```

To transform `a.TestX(3)` we have to do three things:

 1. Find the extension method type name and construct a tree for the invoke expression: `Ext`
 2. Extract the target of the invocation and replace it the tree from the previous step
 3. Add an argument using the extracted target from the previous step

This feels like it should not be as complex as the method suggests. But maybe I am missing something?

Reply:
The main part youre missing is that we  need to reconstruct the whole argumentlist as we can only append, not prepend, as well as that when having extension method calls in the arguments, we have to also redo those (`#95 -121`). 

We then take the all of the identifiers relevant to the invocation and put them in a stack (`# 123-133`) and build the new invocation with it (`#134-145`), adding the new attribute list as well.

Reply:
Check the `SyntaxFactory` class, there is an `ArgumentList` and a `SeparateList` method that should cover the argument transformations that we want.

I don't think that we need to transform any extension method in the arguments. So any `InvocationExpression` that has a parent `InvocationExpression` we do not need to touch.

Consider the following example. The method we are testing is `???.Extension(test, ...)`. The reason we want the normal method invocation instead of the extension method is that it is clearer that this is the method that is tested. The arguments are not part of the test and therefore can remain as is.

```c#
test.Extension(1, 2, other.Extension2());
```